File: EventCountExample/EventCountHybridTopology/PartialCountBolt.cs
using System;
using System.Collections.Generic;
using Microsoft.SCP;
using System.Diagnostics;

namespace EventCountHybridTopology
{
    /// <summary>
    /// Partially count number of messages from EventHubs
    /// </summary>
    public class PartialCountBolt : ISCPBolt
    {
        Context ctx;

        long partialCount = 0L;
        long totalCount = 0L;

        Queue<SCPTuple> tuplesToAck = new Queue<SCPTuple>();

        bool enableAck = false;

        public PartialCountBolt(Context ctx)
        {
            this.ctx = ctx;

            // set input schemas
            Dictionary<string, List<Type>> inputSchema = new Dictionary<string, List<Type>>();
            inputSchema.Add(Constants.DEFAULT_STREAM_ID, new List<Type>() { typeof(string) });

            //Add the Tick tuple Stream in input streams - A tick tuple has only one field of type long
            inputSchema.Add(Constants.SYSTEM_TICK_STREAM_ID, new List<Type>() { typeof(long) });

            // set output schemas
            Dictionary<string, List<Type>> outputSchema = new Dictionary<string, List<Type>>();
            outputSchema.Add(Constants.DEFAULT_STREAM_ID, new List<Type>() { typeof(long) });

            // Declare input and output schemas
            this.ctx.DeclareComponentSchema(new ComponentStreamSchema(inputSchema, outputSchema));

            this.ctx.DeclareCustomizedDeserializer(new CustomizedInteropJSONDeserializer());

            if (Context.Config.pluginConf.ContainsKey(Constants.NONTRANSACTIONAL_ENABLE_ACK))
            {
                enableAck = (bool)(Context.Config.pluginConf[Constants.NONTRANSACTIONAL_ENABLE_ACK]);
            }

            partialCount = 0L;
            totalCount = 0L;
        }

        /// <summary>
        /// The Execute() function will be called, when a new tuple is available.
        /// </summary>
        /// <param name="tuple"></param>
        public void Execute(SCPTuple tuple)
        {
            if (tuple.GetSourceStreamId().Equals(Constants.SYSTEM_TICK_STREAM_ID))
            {
                if (partialCount > 0)
                {
                    Context.Logger.Info("emitting partialCount: " + partialCount +
                        ", totalCount: " + totalCount);
                    if (enableAck)
                    {
                        Context.Logger.Info("tuplesToAck: " + tuplesToAck);
                        this.ctx.Emit(Constants.DEFAULT_STREAM_ID, tuplesToAck, new Values(partialCount));
                        foreach (var tupleToAck in tuplesToAck)
                        {
                            this.ctx.Ack(tupleToAck);
                        }
                        tuplesToAck.Clear();
                    }
                    else
                    {
                        this.ctx.Emit(Constants.DEFAULT_STREAM_ID, new Values(partialCount));
                    }
                    partialCount = 0L;
                }
            }
            else
            {
                partialCount++;
                totalCount++;
                if (enableAck)
                {
                    tuplesToAck.Enqueue(tuple);
                }
            }
        }

        /// <summary>
        ///  Implements of delegate "newSCPPlugin", which is used to create a instance of this spout/bolt
        /// </summary>
        /// <param name="ctx">SCP Context instance</param>
        /// <param name="parms">Parameters to initialize this spout/bolt</param>
        /// <returns></returns>
        public static PartialCountBolt Get(Context ctx, Dictionary<string, Object> parms)
        {
            return new PartialCountBolt(ctx);
        }
    }
}

File: EventCountExample/EventCountTopology/src/main/java/com/microsoft/hdinsight/storm/examples/PartialCountBolt.java
package com.microsoft.hdinsight.storm.examples;

import java.util.Map;
import java.util.LinkedList;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import backtype.storm.Config;
import backtype.storm.Constants;
import backtype.storm.task.OutputCollector;
import backtype.storm.task.TopologyContext;
import backtype.storm.topology.OutputFieldsDeclarer;
import backtype.storm.topology.base.BaseRichBolt;
import backtype.storm.tuple.Fields;
import backtype.storm.tuple.Tuple;
import backtype.storm.tuple.Values;
import backtype.storm.utils.Utils;

/**
 * Partially count number of messages from EventHubs
 */
public class PartialCountBolt extends BaseRichBolt {
  private static final long serialVersionUID = 1L;
  private static final Logger logger = LoggerFactory
      .getLogger(PartialCountBolt.class);
  
  long partialCount = 0L;
  long totalCount = 0L;
  
  private LinkedList<Tuple> tuplesToAck = new LinkedList<Tuple>();
  private OutputCollector collector;
  
  private int tickTupleFreqSecs = 5;
  
  Boolean enableAck = false;
  
  public PartialCountBolt(Boolean enableAck, int tickTupleFreq) {
    this.enableAck = enableAck;
    logger.info("enableAck = " + enableAck);
    this.tickTupleFreqSecs = tickTupleFreq;
    logger.info("tickTupleFreqSecs = " + tickTupleFreqSecs);
  }
  @Override
  public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
    partialCount = 0L;
    this.collector = collector;
  }
  
  @Override
  public void execute(Tuple tuple) {
    //emit partialCount on each TickTuple
    if(isTickTuple(tuple))
    {
      //only emit if partialCount > 0
      if(partialCount > 0) {
        logger.info("emitting intermediate count" + 
            ", partialCount: " + partialCount + 
            ", totalCount: " + totalCount + 
            ", tuplesToAck: " + tuplesToAck.size());
        
        if(enableAck) {
          collector.emit(tuplesToAck, new Values(partialCount));
        } else {
          collector.emit(new Values(partialCount));
        }
        partialCount = 0L;
        
        if(enableAck) {
          for(Tuple tupleToAck : tuplesToAck)
          {
            collector.ack(tupleToAck);
          }
          tuplesToAck.clear();
        }
      }
    }
    else
    {
      partialCount++;
      totalCount++;
      if(enableAck) {
        tuplesToAck.add(tuple);
      }
    }
  }

  @Override
  public Map<String,Object>  getComponentConfiguration() {
    Config conf = new Config();
    //set the TickTuple frequency
    conf.put(Config.TOPOLOGY_TICK_TUPLE_FREQ_SECS, tickTupleFreqSecs);
    return conf;
  }
  
  @Override
  public void declareOutputFields(OutputFieldsDeclarer declarer) {
    declarer.declare(new Fields("partialCount"));
  }
  
  public static boolean isTickTuple(Tuple tuple) {
    return tuple.getSourceComponent().equals(Constants.SYSTEM_COMPONENT_ID)
        && tuple.getSourceStreamId().equals(Constants.SYSTEM_TICK_STREAM_ID);
  }
}